### PR TITLE
BUG: loc dropping levels when df has only one row

### LINKED ISSUE #10521: BUG: inconsisten multi-level indexing when levels are dropped
`DF.loc['A', :, 1]` returns a DataFrame with the full MultiIndex if the second level has more than one entry, and a truncated index if the second level has only one entry (same as, e.g., `DF.loc['A',0,1]`). Is this the intended behavior?

### PR DESCRIPTION
**Pull request by phofl**
- [x] closes #10521
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Originally this case was dispatched to xs, which dispatched to ``_get_loc_level``, which raised if a slice is not ``slice(None, None)`` and went back to our original function. If the ``DataFrame`` contains only one row, ``_get_loc_level`` does not raise and hence we drop levels. 

I added len(self.obj) for performance reasons becasue I am not sure if the any check for slices is faster than supressing the IndexingError in general.

### CODE REVIEW

File: pandas/tests/indexing/multiindex/test_loc.py [hunk 1]
```diff
@@ -695,3 +695,13 @@ def test_loc_getitem_index_differently_ordered_slice_none():
         columns=["a", "b"],
     )
     tm.assert_frame_equal(result, expected)
+
+
+def test_loc_getitem_drops_levels_for_one_row_dataframe():
+    # GH#10521
+    df = DataFrame({"a": ["a"], "b": ["b"], "c": ["a"], "d": 0}).set_index(
```
Comment (jbrockmendel) [inline pandas/tests/indexing/multiindex/test_loc.py:702]: personal preference: to make it super-obvious what is and isn't being tested -- in particular that `set_index` is not being tested -- id prefer this as

```
mi = MultiIndex.from_product(...)
df = DataFrame({"d": 0}, index=mi)
```

Comment (jbrockmendel) [reply pandas/tests/indexing/multiindex/test_loc.py:702]: this would also be easier to follow if the values and the columns didnt overlap

**Comment by jbrockmendel** [comment]
>  If the DataFrame contains only one row, _get_loc_level does not raise and hence we drop levels.

Should the _get_loc_level behavior correct?

**Comment by phofl** [comment]
From reading the code I think this is the intended behavior, but not quite sure.

File: pandas/tests/indexing/multiindex/test_loc.py (hunk 1 above)
Comment (phofl) [reply pandas/tests/indexing/multiindex/test_loc.py]: Thx, simplified it

File: pandas/core/indexing.py [hunk 2]
```diff
@@ -838,8 +838,10 @@ def _getitem_nested_tuple(self, tup: Tuple):
             if self.name != "loc":
                 # This should never be reached, but lets be explicit about it
                 raise ValueError("Too many indices")
-            with suppress(IndexingError):
-                return self._handle_lowerdim_multi_index_axis0(tup)
+            if len(self.obj) > 1 or not any(isinstance(x, slice) for x in tup):
```
Comment (jreback) [inline pandas/core/indexing.py:841]: hmm can we just handle this properly in _handle_lower_multi_index_axis0 ?

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jbrockmendel** [comment]
It looks like jreback and I had different suggestions, but the theme to both is "can this be handled any earlier?"  have you ruled out the lower-level options?

File: pandas/core/indexing.py (hunk 2 above)
Comment (phofl) [reply pandas/core/indexing.py:841]: We could raise an indexingError in there. We can not call xs.

```
            axis = self.axis or 0
            return self._getitem_axis(tup, axis=axis)
```
Would this be preferably to the if condition outside?

**Comment by phofl** [comment]
Sorry tough week at work. We can not call xs, because this would return wrong results. Do you mean mit earlier before reaching this point? Because I understood the suggestions as doing this deeper down in the calling stack

**Comment by jbrockmendel** [comment]
> Do you mean mit earlier before reaching this point? Because I understood the suggestions as doing this deeper down in the calling stack

Deeper in the call stack is ideal, yes.

**Comment by jbrockmendel** [comment]
@phofl are you of the opinion this is the best available approach?  you're pretty well versed in this part of the code and im inclined to trust your opinion.

**Comment by phofl** [comment]
Forgot that one unfortunately, had 3 pretty busy weeks at work, but will be better now.

I am not 100% sure, but I tried this out a bit locally. We have to do this if condition somewhere before dispatching to xs, ``_get_label`` itself is not really an option, would be too ugly. Doing it in ``_handle_lowerdim_multi_index_axis0`` is possible, but would be quite ugly, because the if condition is the same case as raising ``TypeError`` or ``InvalidIndexError``. This would be quite ugly, also I think this is just unnecessary convolution for the other case where ``_handle_lowerdim_multi_index_axis0`` is called.

File: pandas/core/indexing.py (hunk 2 above)
Comment (jreback) [reply pandas/core/indexing.py:845]: why is the `len(self.obj)` needed here? that smells a bit

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/indexing.py (hunk 2 above)
Comment (phofl) [reply pandas/core/indexing.py:845]: Performance reasons, not necessary from a technical standpoint. Will remove it

**Review by jreback — APPROVED** [review]

**Comment by phofl** [comment]
Was a bit hasty. The problem only occurs for Objects with only one row and slices in indexer. Objects with more than one row have to go in there, no matter if the indexer has slices.

**Comment by jreback** [comment]
> Was a bit hasty. The problem only occurs for Objects with only one row and slices in indexer. Objects with more than one row have to go in there, no matter if the indexer has slices.

this is very strange then, these must be hitting a different path that other types, no? maybe something different in the indexing engine itself? its pretty odd to special case like this.

File: pandas/core/indexing.py (hunk 2 above)
Comment (jreback) [reply pandas/core/indexing.py:845]: ok pls add a comment then

**Review by jreback — CHANGES_REQUESTED** [review]
comment and can you merge master

File: pandas/core/indexing.py (hunk 2 above)
Comment (phofl) [reply pandas/core/indexing.py]: Done

**Comment by phofl** [comment]
Looked into this again, the issue is with Series, the dimension of the MultiIndex should be reduced there, #6022

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
cc @jbrockmendel if any commetns.

File: pandas/core/indexing.py [hunk 3]
```diff
@@ -842,8 +842,14 @@ def _getitem_nested_tuple(self, tup: Tuple):
             if self.name != "loc":
                 # This should never be reached, but lets be explicit about it
                 raise ValueError("Too many indices")
-            with suppress(IndexingError):
-                return self._handle_lowerdim_multi_index_axis0(tup)
+            if isinstance(self.obj, ABCSeries) or not any(
```
Comment (jbrockmendel) [inline pandas/core/indexing.py:845]: `self.ndim == 1` is faster than this isinstance check (41 ns vs 297 ns, so not a huge deal either way)

File: pandas/tests/indexing/multiindex/test_loc.py [hunk 4]
```diff
@@ -695,3 +695,17 @@ def test_loc_getitem_index_differently_ordered_slice_none():
...
+    mi = MultiIndex.from_arrays([["x"], ["y"], ["z"]], names=["a", "b", "c"])
+    df = DataFrame({"d": [0]}, index=mi)
+    expected = df.copy()
+    result = df.loc["x", :, "z"]
+    tm.assert_frame_equal(result, expected)
+
+    ser = Series([0], index=mi)
+    result = ser.loc["x", :, "z"]
+    expected = Series([0], index=Index(["y"], name="b"))
+    tm.assert_series_equal(result, expected)
```
Comment (jbrockmendel) [inline pandas/tests/indexing/multiindex/test_loc.py:711]: big picture, why isnt `ser_expected = frame_expected["d"]`?

Comment (phofl) [reply pandas/tests/indexing/multiindex/test_loc.py:711]: Series drops levels of MultiIndex (for example #6022), while DataFrame should keep them. The current behavior is, that DataFrame keeps them all the times except when it has one row.

File: pandas/core/indexing.py (hunk 3 above)
Comment (phofl) [reply pandas/core/indexing.py]: Is better to read too :)

**Comment by jreback** [comment]
thanks @phofl

File: pandas/tests/indexing/multiindex/test_loc.py (hunk 4 above)
Comment (jbrockmendel) [reply pandas/tests/indexing/multiindex/test_loc.py:711]: is https://github.com/pandas-dev/pandas/pull/38150#issuecomment-752233136 saying that you plant to change the Series behavior to match DataFrame?

Comment (phofl) [reply pandas/tests/indexing/multiindex/test_loc.py:711]: No, the comment was meant to clarify why we have to check for ndim==1

File: pandas/core/indexing.py [hunk 5]
```diff
@@ -842,8 +842,12 @@ def _getitem_nested_tuple(self, tup: Tuple):
             if self.name != "loc":
                 # This should never be reached, but lets be explicit about it
                 raise ValueError("Too many indices")
-            with suppress(IndexingError):
-                return self._handle_lowerdim_multi_index_axis0(tup)
+            if self.ndim == 1 or not any(isinstance(x, slice) for x in tup):
```
Comment (jbrockmendel) [inline pandas/core/indexing.py:845]: @phofl the more i look at this the weirder it seems to be treating Series/DataFrame differently.  Are you clear on if there's a reason for this vs just historical coincidence?

Comment (phofl) [reply pandas/core/indexing.py:845]: @jbrockmendel 

sorry for the slow reply. I am pretty sure that this is for historical reasons, but would have to take a closer look.

I am currently on vacation until mid July. Can get back to you afterwards, will also have more time for pandas then again.

Comment (jbrockmendel) [reply pandas/core/indexing.py:845]: no worries, enjoy your vacation

**Pull request merged**